fix(ssi): split default label selectors

betterengineering opened the pull request:
### What does this PR do?
This commit splits the default label selectors for SSI. Other webhooks should not change behavior based on if instrumentation is enabled or not. Similarly, our webhook should not receive requests for disabled namespaces or other SSI specific use cases.

### Motivation
We had a customer case where the `config` webhook was applying to namespaces outside of the configured namespace for `autoinstrumentation`. The lead the customer to reach out because they were getting mutation outside of their configured namespaces. However, they were not getting mutation for SSI, they were only getting mutation for the config webhook.

Other webhooks should not be relying on `apm_config.instrumentation.enabled` or `apm_config.instrumentation.enabled_namespaces` as these keys are used to control Single Step Instrumentation enablement. In the customer case, these values being truthy lead the `config` webhook to inject every pod in every namespace except the default 🙀 .

By splitting the label selector logic and removing Single Step Instrumentation related config from the default, we can resolve the customer issue:
* `autoinstrumentation` will continue to filter pods as expected and have a slight efficency gain that we can make better over time
* `config` will only mutate pods with either the `dmission.datadoghq.com/enabled` label set to `true` or have `admission_controller.mutate_unlabelled` set to `true` and will still respect the default disabled namespaces.

### Describe how you validated your changes
I've added more unit tests here. The main test I've been using is through injector-dev:

<details><summary>duplicate-workloads.yaml</summary>

```yaml
helm:
  apps:
    - name: injection-expected
      namespace: apptest4
      values:
        env:
          - name: DD_TRACE_DEBUG
            value: "true"
          - name: DD_APM_INSTRUMENTATION_DEBUG
            value: "true"
        image:
          repository: registry.ddbuild.io/ci/injector-dev/python
          tag: 2cd78ded
        podLabels:
          language: python
          tags.datadoghq.com/env: local
        service:
          port: "8080"
    - name: injection-not-expected
      namespace: testapp2
      values:
        env:
          - name: DD_TRACE_DEBUG
            value: "true"
          - name: DD_APM_INSTRUMENTATION_DEBUG
            value: "true"
        image:
          repository: registry.ddbuild.io/ci/injector-dev/python
          tag: 2cd78ded
        podLabels:
          language: python
          tags.datadoghq.com/env: local
        service:
          port: "8080"
  versions:
    agent: 7.72.2
    cluster_agent:
      version: 7.72.2
      build: {}
    injector: 0.52.0
  config:
    datadog:
      apm:
        instrumentation:
          enabled: true
          enabledNamespaces:
            - "apptest4"
          libVersions:
            python: "3"

```

</details> 

This will mutate both workloads with the `config` webhook and will mutate `apptest4` to use SSI:
```bash
injector-dev apply -f duplicate-webhooks.yaml
```

Building my branch will only mutate `apptest4` for SSI:
```bash
injector-dev apply -f duplicate-webhooks.yaml --build
```

### Additional Notes
Open to suggestions! This felt like a clean path forward.

betterengineering committed:
fix(ssi): split default label selectors

This commit splits the default label selectors for SSI. Other webhooks
should not change behavior based on if instrumentation is enabled or
not. Similarly, our webhook should not receive requests for disabled
namespaces.

Mathew-Estafanous reviewed (approved):
Decoupling apm label selector config from the default should be a quick win for our customers - especially because we've already had several support cases come our way.

**Note:** This PR does change the default behaviour of our `config` and `tag` mutation webhooks. Currently an 'opt-out' label selector is generated if `mutate_unlabelled` OR `"apm_config.instrumentation.enabled` is enabled. Now that we've decoupled that condition, config socket injection may stop for customers who've unknowingly been relying on this bug.

We should communicate this change in behaviour depending on how many customers may be impacted

betterengineering commented on pkg/clusteragent/admission/mutate/autoinstrumentation/label_selectors.go:
```diff
@@ -0,0 +1,100 @@
+
+// Get returns the namespace and object selectors based on the configuration and the optional useNamespaceSelector.
+func (ls *LabelSelectors) Get(useNamespaceSelector bool) (*metav1.LabelSelector, *metav1.LabelSelector) {
+	var objectSelector *metav1.LabelSelector
```
Ok, I tried to refactor this to something like:
```go
	return common.DefaultLabelSelectors(useNamespaceSelector, common.LabelSelectorsConfig{
		ExcludeNamespaces: ls.config.DisabledNamespaces,
    })
```

But the problem is I also am checking if APM is enabled down below which the default one is no longer doing:
```go
if ls.config.Enabled || ls.config.MutateUnlabelled {
```

Maybe I could pass in an `enabled` option? But ultimately I think it would be good to divorce these implementations. SSI could go even further with this, given we could generate a label selector based on the targeting rules the user provides as a further optimization. What do you think?

betterengineering committed:
Fix linter

betterengineering committed:
Add release note

betterengineering committed:
Updated based on Dom's suggestions

domalessi reviewed (approved):
Just a few formatting tweaks, but looks great!

betterengineering committed:
Update releasenotes-dca/notes/split-default-label-selectors-7e48728f7aa67a7d.yaml

Co-authored-by: domalessi <111786334+domalessi@users.noreply.github.com>

betterengineering commented:
/merge

betterengineering committed:
chore(agent_sidecar): ensure new env vars are prepended

This commit ensures that newly added environment variables are prepended
in the agent sidecar webhook. Prepend rather than append the new
variables so that they precede the previous ones in the final list,
allowing them to be referenced in other environment variables downstream.

betterengineering committed:
Revert "chore(agent_sidecar): ensure new env vars are prepended"

This reverts commit 0a58401ea6476a8f6e2bd8b3f71bb2fbd9792851.

betterengineering committed:
chore(ssi): add enabled label to test app

This commit adds the enabled label to the test app so that it will now
receive mutations from the config webhook. The reason this worked before
is because the config webhook applied when apm.instrumentation.enabled
was true.
